bpf: Replace join_ep.sh with Golang code

joestringer opened the pull request:
Revive #2411.

Discussions on two recent issues have proposed optimizing out the
compile step for particular aspects of regeneration, by creating a new
map with the desired contents in the canonical location for the map,
then invoking the loader directly to update references to the new
versions of the map without recompiling the datapath code: #3897, #5303.

We have also recently been investigating exporting metrics for
compilation times to optimize the performance of Cilium: #5329.

The current join_ep.sh script doesn't support only compiling or only
loading, and furthermore unless we split this into two different bash
scripts, it will be difficult for proper measurement of the compilation
time vs. load time.

To provide a path to support the above optimization, and also to help
with measuring different parts of endpoint regeneration, refactor the
join_ep.sh script to instead directly invoke each foreign binary
directly from Golang code.

Related: #1600, #1711, #2411, #3897, #5303, #5329

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5408)
<!-- Reviewable:end -->

joestringer committed:
endpoint: Export Logger accessor

An upcoming commit desires to access the endpoint log from a package
which is called from pkg/endpoint. Set up for this by exporting the
function to fetch the logger.

While we're at it, update the name to "Logger" to be more consistent
with regular golang style:

  https://golang.org/doc/effective_go.html#Getters

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
endpoint: Shift join_ep.sh access to loader package

Factor out the shelling out to compile & load programs into its own
package, to reduce the BPF-related dependency code in pkg/endpoint.

Signed-off-by: Joe Stringer <joe@covalent.io>

aanm commented on pkg/datapath/loader/loader.go:
```diff
@@ -0,0 +1,103 @@
+
+const (
+	// ExecTimeout is the execution timeout to use in join_ep.sh executions
+	ExecTimeout = 300 * time.Second
```
Can this be private?

aanm commented on pkg/datapath/loader/loader.go:
```diff
@@ -0,0 +1,103 @@
+// Expects the caller to have created the directory at the path ep.StateDir().
+func CompileAndLoad(ep endpoint, cfg config) error {
+	if ep == nil {
+		log.Fatalf("LoadBPF() doesn't support non-endpoint load")
```
I would return an error to the caller instead of fatal.

joestringer replied:
This is a developer bug.

ianvernon commented on pkg/endpoint/bpf.go:
```diff
@@ -254,53 +248,38 @@ func hashHeaderfile(hashWriter hash.Hash, filepath string) (hash.Hash, error) {
-	keys     []*lxcmap.EndpointKey
-	value    *lxcmap.EndpointInfo
-	ifName   string
+	// revision is used by the endpoint regeneration code to determine
```
Nit: can you move `epInfoCache` code to a separate file?

ianvernon reviewed (approved):
LGTM for endpoint, CI, CLI.

aanm commented on pkg/datapath/loader/compile.go:
```diff
@@ -0,0 +1,182 @@
+// irPath determines a temporary path on the filesystem to store the compiled
+// intermediate representation object for the program.
+//
+// TODO: Remove this and just pipe the 'clang' and 'llc' programs together.
```
If you want to have some fun: https://stackoverflow.com/questions/10781516/how-to-pipe-several-commands-in-go

aanm commented on pkg/datapath/loader/compile.go:
```diff
@@ -0,0 +1,182 @@
+// Link the specified program from the specified path to the intermediate
+// representation, to the output specified in the prog's info.
+func Link(ctx context.Context, prog *progInfo, dir *directoryInfo, debug bool) error {
+	args := make([]string, 0, 8)
```
Where does the 8 comes from?

joestringer replied:
We typically use less than 8 ld flags. I looked at how many we used, and picked a number. It doesn't really matter, and the capacity will expand if the number isn't right, but we also don't change these very often.

ianvernon replied:
@joestringer can you add a comment indicating as such in the code?

aanm commented on pkg/datapath/loader/compile.go:
```diff
@@ -0,0 +1,182 @@
+
+// compile and link a program.
+func compile(ctx context.Context, prog *progInfo, dir *directoryInfo, debug bool) error {
+	args := make([]string, 0, 16)
```
why 16?

joestringer replied:
Similar to ldflags, I looked at how many cflags we use and it's typically less than 16.

aanm commented on pkg/datapath/loader/compile.go:
```diff
@@ -0,0 +1,182 @@
+		scanner := bufio.NewScanner(bytes.NewReader(out))
+		return fmt.Errorf("Failed to compile %q: %s: %q", prog.Output, err, scanner.Text())
+	}
+	defer os.Remove(irPath(prog, dir))
```
If line 165 fails, does it mean the path won't be clean up?

joestringer replied:
If line 165 fails, then the path won't be created in the first place. Clang creates it, we pass it into llc, then clean up afterwards.

aanm commented on pkg/datapath/loader/compile.go:
```diff
@@ -0,0 +1,182 @@
+
+// Link the specified program from the specified path to the intermediate
+// representation, to the output specified in the prog's info.
+func Link(ctx context.Context, prog *progInfo, dir *directoryInfo, debug bool) error {
```
Does it need to be public?

joestringer replied:
Nope, can hide.

aanm commented on pkg/datapath/loader/compile.go:
```diff
@@ -0,0 +1,182 @@
+
+	switch prog.OutputType {
+	case outputObject:
+		return Link(ctx, prog, dir, debug)
```
what about "`return Zelda`"? :smile: Can you rename the function to `LinkProg`?

aanm commented on pkg/datapath/loader/loader.go:
```diff
@@ -54,34 +48,42 @@ type endpoint interface {
+// * Assembly
+// * Object compiled with debug symbols
+func compileDatapath(ep endpoint, dirs *directoryInfo, debug bool) error {
+	// TODO: Consider logging kernel/clang versions here too
```
We have them when we start cilium-agent.

joestringer replied:
Yep, would just need a little refactoring to make available. Basically this is just a comment saying that when we ran `join_ep.sh` before, we would always print these first, inline, in the logs. We also do this on agent startup, but it's possible to lose the logs from when the daemon first started.

joestringer committed:
endpoint: Split endpoint cache into cache.go

During compilation, it's useful to grab a cache of the relevant endpoint
info that can be used by the next layer of code to generate the BPF
datapath, then release the lock on the endpoint to allow other work to
continue. This cache facilitates this, but may end up holding
stale information about an endpoint which may be subsequently deleted.

This patch just refactors the code related to this cache into its own
file for better code organisation.

Signed-off-by: Joe Stringer <joe@covalent.io>

aanm commented on pkg/datapath/loader/netlink.go:
```diff
@@ -0,0 +1,106 @@
+		QdiscType:  "clsact",
+	}
+
+	// Ignore failure on replacing clsact, means we replaced it.
```
This sentence doesn't make sense to me. Even if QdiscReplace returns an error it meas we have replace it? :thinking:

joestringer replied:
Let me double-check this. We don't check the error today, probably because it can return an error if there's already a qdisc there (for example, because the qdisc we want to install already exists).

Furthermore, this should succeed the first time because *we* create the link. So it should only end up in a bad state if someone's really messing around with the devices.

joestringer replied:
OK, I traced this down to the netlink library inserting an extra `TCA_OPTIONS` option to the command with empty contents, which is unnecessary and messes with the replacement of the qdisc. :/

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -633,22 +633,19 @@ func (e *Endpoint) regenerateBPF(owner Owner, epdir string, regenContext *Regene
 		start := time.Now()
 		// Compile and install BPF programs for this endpoint
-		err = e.runInit(libdir, rundir, epdir, epInfoCache.ifName, debug)
+		err = loader.CompileAndLoad(epInfoCache, owner)
```
[0] Here

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -254,53 +248,38 @@ func hashHeaderfile(hashWriter hash.Hash, filepath string) (hash.Hash, error) {
-
-	scopedLog := e.getLogger()
-
-	ctx, cancel := context.WithTimeout(context.Background(), ExecTimeout)
```
This context was created before doing any Command execution. Can you do the same before calling the CompileandLoad in [0]? This way you are reusing a single ctx for everything call made. Be aware you will need to check for context.DeadlineExceeded before executing any preceding "expensive" function call.

joestringer replied:
Is there a link somewhere or something that explains why you need to check `context.DeadlineExceeded` before executing subsequent calls? Won't the call just return an error that we can pass back?

joestringer replied:
Ah, apparently it won't. Okay. I need to fix this up across the PR.

joestringer replied:
And done.

joestringer committed:
loader: Add benchmarks

We skip adding unit tests here to avoid introducing unit-test
dependencies on stuff like clang, llc, cilium-map-migrate, etc.

Signed-off-by: Joe Stringer <joe@covalent.io>

vadorovsky commented on pkg/datapath/loader/compile.go:
```diff
@@ -0,0 +1,188 @@
+	return []string{
+		fmt.Sprintf("-filetype=%s", prog.OutputType),
+		irPath(prog, dir),
+		"-o", fmt.Sprintf("%s/%s", dir.Output, prog.Output),
```
`path.Join`?

vadorovsky commented on pkg/datapath/loader/compile.go:
```diff
@@ -0,0 +1,188 @@
+	var output string
+
+	if prog.OutputType == outputSource {
+		output = fmt.Sprintf("%s/%s", dir.Output, prog.Output)
```
`path.Join`?

vadorovsky commented on pkg/datapath/loader/loader.go:
```diff
@@ -0,0 +1,114 @@
+	}
+
+	// Replace the current program
+	objPath := fmt.Sprintf("%s/%s", dirs.Output, endpointObj)
```
`path.Join`?

vadorovsky commented on pkg/datapath/loader/netlink.go:
```diff
@@ -0,0 +1,117 @@
+	// FIXME: Replace cilium-map-migrate with Golang map migration
+	out, err := exec.CommandContext(ctx, "cilium-map-migrate", "-s", objPath).CombinedOutput()
+	if ctx.Err() == context.DeadlineExceeded {
+		err = fmt.Errorf("Command execution failed: Timeout")
```
I see that this if statement is repeated couple of times. Maybe we should have some helper function like:

```
func CommandWithTimeout(ctx context.Context, args ...string) error {
    out, err := exec.CommandContext(ctx, args).CombinedOutput()
    if ctx.Err() == context.DeadlineExceeded {
        return fmt.Errorf("Command execution failed: Timeout")
    }
    return err
}
```

joestringer replied:
Having looked into this, I agree, but I'd rather tackle it in a separate PR since there's places around in `daemon/` and potentially other bits that could make use of such a generic helper.

joestringer replied:
PR: #5439

vadorovsky replied:
Nice, thanks!

vadorovsky commented on pkg/endpoint/cache.go:
```diff
@@ -0,0 +1,96 @@
+// BPF endpoints map
+// Must only be called if init() succeeded.
+func (ep *epInfoCache) GetBPFValue() (*lxcmap.EndpointInfo, error) {
+	return ep.value, nil
```
If we are never going to return any `err` there, why do we need the second value? Is it for implementing some interface?

joestringer replied:
Yes, this is satisfying `lxcmap.EndpointFrontend`.

joestringer committed:
loader: Pipe compiler into linker

Remove the temporary file that was previously being used between the
compiler and linker, and just pipe the two processes together.

Signed-off-by: Joe Stringer <joe@covalent.io>

vadorovsky commented on pkg/datapath/loader/compile.go:
```diff
@@ -0,0 +1,188 @@
+}
+
+var (
+	compiler       = "clang"
```
I don't want to block that PR with that, but maybe in future we should consider making that configurable. Users can have multiple versions of clang installed and sometimes in stable distros (RHEL, SLES etc.) the older versions are proposed as default ones. It would be nice to provide a possibility of picking a specific version (i.e. `clang6`).

vadorovsky commented on pkg/datapath/loader/compile.go:
```diff
@@ -0,0 +1,188 @@
+func linkProg(ctx context.Context, prog *progInfo, dir *directoryInfo, debug bool) error {
+	args := make([]string, 0, 8)
+	if debug {
+		args = append(args, "-mattr=dwarfris")
```
Should we also add the `-debug` option of llc there? Or are we fine only with BTF debug info?

joestringer replied:
I don't see a `-debug` option in my llc. Either way, we can always consider further extensions. This PR should be largely just a refactor without functional changes.

vadorovsky reviewed (changes requested):
Some nits and questions

joestringer committed:
loader: replace join endpoint script with loader

We are moving away from invoking scripts to a more programmatic
approach.  This patch still invokes the commands but wraps it all in Go.
This will make it easier to fully replace the init / setup script.

From discussions with Daniel, the next thing is to get the iproute2 code
accessible via cgo or swig. If something like that works we can call tc,
bpf and related functions directly from Go. Then the needed calls are
compiled in and the iproute2 version dependency can be dropped.

Maybe too much work, but it would be nice if we also had a option to use
LLVM apis instead of clang and llc execs for the container case where we
can decide which version is being shipped.

Related: #1711 (evaluate vishvananda/netlink for replacing iproute2 usage)
Related: #1600 (Go native interaction with the kernel)

Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>
Co-authored-by: Joe Stringer <joe@covalent.io>
Signed-off-by: Joe Stringer <joe@covalent.io>

aanm reviewed (approved):
LGTM Do you have any benchmarks WRT this approach and the previous one to compile and load bpf?
